### ISSUE TITLE
Missing references for case class if search starts from different file

### ISSUE THREAD

**Issue by tgodzik** [scalameta/metals#1553]
**Describe the bug**
If we look for references of a case classes's synthetic companion object from a file that doesn't contain the definition, we will not find all the correct references.

This is due to the fact that we only look for alternative references inside the current file. We need to first find semanticDB for the definition of a symbol and then use that.

**To Reproduce**

Add test case to ReferencesLspSuite:
```scala
  check(
    "companion",
    """|/a/src/main/scala/a/Main.scala
       |package a
       |case class <<Main>>(name : String){
       |  val appl = <<Main>>.apply("")
       |}
       |/a/src/main/scala/a/Other.scala
       |package a
       |object Other{
       | val appl = <<M@@ain>>.apply("")
       |}
       |""".stripMargin
  )
```

**Expected behavior**
All references to the case classes are returned.

**Installation:**
- Operating system: macOS/Windows/Linux
- Editor: Visual Studio Code/Atom/Vim/Sublime/Emacs
- Metals version: v0.8.3

**Additional context**

**Search terms**
references rename case class

**Comment by tgodzik** [comment]
Should be possible to reuse method from `ImplementationProvider`, which should msot likely be moved to definitionProvider:

```scala
  private def findSemanticDbForSymbol(symbol: String): Option[TextDocument] = {
    for {
      symbolDefinition <- index.definition(MSymbol(symbol))
      document <- findSemanticdb(symbolDefinition.path)
    } yield {
      document
    }
  }
```

**Comment by faizk** [comment]
Ah, someone finally found this.

I ran into the exact same thing problem a couple of months ago, but in my case it as an `unapply` that wasn't getting found:

```scala
// /a/src/main/scala/a/Stuff.scala
package a
sealed trait Stuff
case class X(n: Int) extends Stuff // looking for references to this,
```
```scala
// /a/src/main/scala/a/Main.scala
package a
object Main {
   def n(stuff: Stuff): Option[Int] = stuff match {
     case X(n) => Some(n) // this one won't be found.
     case _ => None
   }
}
```
Before I could figure out how to appropriately express this as a failing unit-test, I mucked around with `ReferenceProvider.scala` trying to find out how synthetics are getting missed - in the process wrote this hack just to test the theory that synthetics across the file border weren't getting found.
 
https://github.com/faizk/metals/commit/e837a24a32997ae858ace5128969e59b931b647e

Besides being hacky, it's likely inefficient (it does extra lookups for synthetics when doing lookups for classes, `case` or not), but I used it to unblock my work, and wanted to send a proper pull-request after figuring out which provider it belongs to, how best to test, etc,

**Comment by tgodzik** [comment]
@faizk References issues can now be reported in a similar way in the ReferenceLspSuite if you find any more. It wasn't as simple before.

The issue here is when we generate 'alternatives':
```
val alternatives = referenceAlternatives(doc, occurrence)
```
The doc value is the local one, we need to use `findSemanticDbForSymbol` to get the right one.
It most likely can be lazy val in case it's not something we need an alternative for.

You are welcome to create a PR for it! It should not be too complex.

**Comment by faizk** [comment]
Thanks, I'll give it a shot.

**Comment by tgodzik** [comment]
> Thanks, I'll give it a shot.

We're planning to put some work into it, since we are having a new contributor coming into the project and this could a good starting issue.

However, if you really want to take a look at it and have some time to do it let me know!

**Comment by faizk** [comment]
I sort of dived into it over the weekend - seems like a lot of fun so far! - which is why I'm hoping to get a PR out soon; I might reach out just to make sure I'm headed in the right direction. I think if it takes me longer than a couple of days then I'll fail-fast and let you know!

**Comment by tgodzik** [comment]
> I sort of dived into it over the weekend - seems like a lot of fun so far! - which is why I'm hoping to get a PR out soon; I might reach out just to make sure I'm headed in the right direction. I think if it takes me longer than a couple of days then I'll fail-fast and let you know!

Cool, if you already divided into it then you should continue. We'll prioritize something else in the meantime. Let us know if you need any help!

**Comment by faizk** [comment]
I made some progress over the last couple of days - thanks for the pointer with `referenceAlternatives`! I'm think I'm close to a PR, thanks for your patience. What's a good place to talk about unit-tests (structuring/refactoring them)? The gitter channel sounds more like it's for general usage.

**Comment by tgodzik** [comment]
I think you can discuss that on the discord channel, which has a bit more rooms and we monitor it with the same frequency. You can find it in the bottom of the page on https://scalameta.org/metals/

**Comment by faizk** [comment]
Note that even this fails: 
(it's the same as the original test, but even though all references live in the same compilation unit `a/Main.scala`, still doesn't find all references:)

```scala
  check(
    "companion-same-unit",
    """|/a/src/main/scala/a/Main.scala
       |package a
       |case class <<Main>>(name : String){
       |  val appl = <<Main>>.apply("")
       |}
       |
       |object Other{
       | val appl = <<M@@ain>>.apply("")
       |}
       |""".stripMargin
  )
```
fails with:
```
=> Diff (- obtained, + expected)
 package a
-case class Main(name : String){
+case class <<Main>>(name : String){
   val appl = <<Main>>.apply("")
```


Just to be clear about the nature of the bug:
To succeed, the references to synthetics (`apply`/`unapply`) need to be in the same compilation unit, AND the query needs to start on the declaration. So this is the only variant that does succeed:

```scala
  check(
    "companion-from-decl-same-unit",
    """|/a/src/main/scala/a/Main.scala
       |package a
       |case class <<M@@ain>>(name : String){
       |  val appl = <<Main>>.apply("")
       |}
       |object Other{
       | val appl = <<Main>>.apply("")
       |}
       |""".stripMargin
  )
```

Here's the tests I added on top of v0.8.4 - a bit verbose 
but just want to make sure I've not missing anything subtle.

https://gist.github.com/faizk/3168a3ed4464b112762149faa6804bc0

**Comment by tgodzik** [comment]
@faizk looks good! This is actually 2 bugs here:
- we only find reference alternatives inside the same file
- we don't create proper reference alternatives if something is such alternative. In particular for synthetic `Foo.unapply().` we don't generate `Foo#` and `Foo.` 

I perfectly happy with first fixing the first one and following up with the second one later.

Great work tracking that other bug!

**Comment by faizk** [comment]
> we don't create proper reference alternatives if something is such alternative. In particular for synthetic `Foo.unapply().` we don't generate `Foo#` and `Foo`.

OK, as I feared, that's during indexing. I was able to see how in the semanticdb of the file that's in the same package-space as the type `Foo` but only references synthetics is missing exactly those (`Foo#` and `Foo`).

I'm guessing something around:

```scala
// ReferenceProvider.scala lines 61-66

      d.synthetics.foreach { synthetic =>
        Synthetics.foreachSymbol(synthetic) { sym =>
          bloom.put(sym)
          Synthetics.Continue
        }
      }
```

I tried a hack where I coax in an extra `bloom.put` but that doesn't prove anything new - I'm going to checkout semanticdb to take a better look. But you're right, these are 2 different bugs and I'll split it up like you suggest.

**Comment by tgodzik** [comment]
> OK, as I feared, that's during indexing. I was able to see how in the semanticdb of the file that's in the same package-space as the type `Foo` but only references synthetics is missing exactly those (`Foo#` and `Foo`).

It should actually be possible to do an inverse operation. For each alternative generate the possible base symbol if the definition of this alternative doesn't exists and search for the semanticDB for it.

**Comment by faizk** [comment]
>     * we only find reference alternatives inside the same file

The PR that address this point:
https://github.com/scalameta/metals/pull/1727
